pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1825,63 +2073,207 @@ func (og *operationGenerator) nodeExpandVolume(
+		} else {
+			return og.legacyCallNodeExpandOnPlugin(resizeOp)
+		}
+	}
+	return true, nil
+}
 
-			if volumeToMount.VolumeSpec.ReadOnly {
-				simpleMsg, detailedMsg := volumeToMount.GenerateMsg("MountVolume.NodeExpandVolume failed", "requested read-only file system")
-				klog.Warningf(detailedMsg)
-				og.recorder.Eventf(volumeToMount.Pod, v1.EventTypeWarning, kevents.FileSystemResizeFailed, simpleMsg)
-				og.recorder.Eventf(pvc, v1.EventTypeWarning, kevents.FileSystemResizeFailed, simpleMsg)
-				return true, nil
-			}
-			rsOpts.VolumeSpec = volumeToMount.VolumeSpec
-			rsOpts.NewSize = pvSpecCap
-			rsOpts.OldSize = pvcStatusCap
-			resizeDone, resizeErr := expandableVolumePlugin.NodeExpand(rsOpts)
-			if resizeErr != nil {
-				// if driver returned FailedPrecondition error that means
-				// volume expansion should not be retried on this node but
-				// expansion operation should not block mounting
-				if volumetypes.IsFailedPreconditionError(resizeErr) {
-					actualStateOfWorld.MarkForInUseExpansionError(volumeToMount.VolumeName)
-					klog.Errorf(volumeToMount.GenerateErrorDetailed("MountVolume.NodeExapndVolume failed with %v", resizeErr).Error())
-					return true, nil
+// callNodeExpandOnPlugin is newer version of calling node expansion on plugins, which does support
+// recovery from volume expansion failure.
+func (og *operationGenerator) callNodeExpandOnPlugin(resizeOp nodeResizeOperationOpts) inTreeResizeResponse {
+	pvc := resizeOp.pvc
+	pv := resizeOp.pv
+	volumeToMount := resizeOp.vmt
+	rsOpts := resizeOp.pluginResizeOpts
+	actualStateOfWorld := resizeOp.actualStateOfWorld
+	expandableVolumePlugin := resizeOp.volumePlugin
+
+	var err error
+	pvcStatusCap := pvc.Status.Capacity[v1.ResourceStorage]
+	pvSpecCap := pv.Spec.Capacity[v1.ResourceStorage]
+
+	resizeResponse := inTreeResizeResponse{
+		pvc: pvc,
+		pv:  pv,
+	}
+
+	if permitNodeExpansion(pvc, pv) {
+		// File system resize was requested, proceed
+		klog.V(4).InfoS(volumeToMount.GenerateMsgDetailed("MountVolume.NodeExpandVolume entering", fmt.Sprintf("DevicePath %q", volumeToMount.DevicePath)), "pod", klog.KObj(volumeToMount.Pod))
+
+		rsOpts.VolumeSpec = volumeToMount.VolumeSpec
+		rsOpts.NewSize = pvSpecCap
+		rsOpts.OldSize = pvcStatusCap
+		pvc, err = util.MarkNodeExpansionInProgress(pvc, og.kubeClient)
+
+		if err != nil {
+			msg := volumeToMount.GenerateErrorDetailed("MountVolume.NodeExpandVolume failed to mark node expansion in progress: %v", err)
+			klog.Errorf(msg.Error())
+			resizeResponse.err = msg
+			return resizeResponse
+		}
+
+		resizeDone, resizeErr := expandableVolumePlugin.NodeExpand(rsOpts)
+		resizeResponse.resizeCalled = true
+
+		if resizeErr != nil {
+			if volumetypes.IsOperationFinishedError(resizeErr) {
+				var markFailedError error
+				pvc, markFailedError = util.MarkNodeExpansionFailed(pvc, og.kubeClient)
+				// update the pvc with node expansion object
+				resizeResponse.pvc = pvc
+				if markFailedError != nil {
```

Comment:
should there be `resizeResponse.assumeResizeOpAsFinished = true`?

Reply:
fixed.